### PR TITLE
adding login and logout

### PR DESCRIPTION
**Pull request by jarlah**
Trying to get familiar with vue. Followed a tut on https://scotch.io/tutorials/handling-authentication-in-vue-using-vuex

### CODE REVIEW

File: src/store.ts [hunk 1]
```diff
@@ -1,16 +1,72 @@
...
+                axios.defaults.headers.common.Authorization = token;
+                commit('auth_success', {token, username});
+                resolve(resp);
+              })
+              .catch((err) => {
+                commit('auth_error');
+                localStorage.removeItem('token');
+                reject(err);
+              });
+        }, 2000);
```
Comment (jarlah) [inline src/store.ts]: just a test to see how the login form looked while waiting for response. I made the button disabled in login page when isLoading is true. By waiting 2secs i get a feeling for how it would look and feel in the real world. Should be removed before merge?

**Comment by jarlah** [comment]
Actually I think the linter settings is pretty much annoying. And using typescript seems to be pretty useless. However, its good to have the option.

File: src/views/Login.vue [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        },
+        isError() {
+            return this.$store.getters.isError;
+        },
+    },
+    data() {
+        return {
+            username: '',
+            password: '',
+            error: undefined,
```
Comment (jarlah) [inline src/views/Login.vue]: Currently not using the error for anything. The template uses isError which comes from global state. I could maybe Just dispatch the error to the store.

I feel this model is quite different from react and redux. And i have to get used to it.

Comment (jarlah) [reply src/views/Login.vue]: Could i maybe handle all error cases in the global store? And drop the catch block below?

**Review by jarlah — COMMENTED** [review]
I have some uncertanties regarding a couple of changes. But overall i think its acceptable.

**Comment by jarlah** [comment]
Investigating http://georgekinsman.com/programming/2018/04/19/Typed-State-Management-With-Vuex.html

...

**Comment by jarlah** [comment]
The typescript compiler is fishy. I have seen this behavior before. It doesnt fail typescript compilation if obvious mistakes has been done. I think however that this has nothing to do with vue

**Comment by pingwiniasty** [comment]
Hi, I will check it tomorrow. Today I have been away whole day

**Comment by jarlah** [comment]
> Hi, I will check it tomorrow. Today I have been away whole day

@pingwiniasty  i finally found out what was wrong with typescript compilation. The problem is that the .vue compiler is not able to typecheck typescript. So therefore the solution is to do like this:

<script src="./Login.ts"></script>

And huhey we get lots of compiler errors. I love compiler errors.

Only problem we might encounter is third party libraries. But the shims we need to make for those is a cheap cost.

**Comment by pingwiniasty** [comment]
Are you sure you want to split css from components to separate files?

**Comment by jarlah** [comment]
Tldr i will move css back to vue files 

@pingwiniasty not necessarily. I just had to split the code to have full control of its compilation. I can move back the css into the vue file. I was really Just testing. Looked at angular yesterday. I died multiple times by watching the terrible file structure. Angular is also kind of view first, but you are almost forced to split into multiple files. 

In addition there is nothing stopping us from using Vue.extend instead of class components, but class components give more control of where fields come from instead of relying on unbounded this in vue extend.

**Comment by jarlah** [comment]
@pingwiniasty Done. Moved css back into vue files.

**Comment by jarlah** [comment]
btw @pingwiniasty I converted the hello world component to es6 style with plain object export. To show that If we need to, we could use whatever suits the purpose in the vue files. But if you are going to use plain object vue components, you must create a bunch of tests to prove that it works. Because it will lack type safety. I will also create tests for typescript components, to prove that those also work. But they do not need as much as testing as plain js

**Comment by jarlah** [comment]
We could discuss the use of external files vs embedded. When you have a method that are accessing any fields like this name = "dkdk" or using imported typescript code like the authentication module, i would prefer typescript. If you only have a component where it takes input props and does some confitional view logic in the template we could javascript.

**Comment by pingwiniasty** [comment]
Mixing 2 types of code style are not good option. Vue 3.x will improve TypeScript support, and we would be prepared for this.

**Comment by jarlah** [comment]
Ok. @pingwiniasty the hello world component is just a dummy. So ill just leave it as it is now. It will be removed anyway together with the rest of the home module.

If the rest looks good, except for hello world component, you are free to approve and merge.

**Comment by jarlah** [comment]
Wait. Ill try Vetur extension for visual code. Too see if it can work satisfactory with SFC's. (single file etc..)

**Comment by jarlah** [comment]
heh. I really wonder why i began to think SFC's do not work with with TS... i think i have listened to fake news, or old news, or something. I have heard that typescript doesnt play well with vue. but it does! I may have just mistyped <script type="ts"> instead of <script lang="ts"> .. now the code will light up as red neon lights if i do something fishy :)

**Comment by jarlah** [comment]
@pingwiniasty  I finally found out that ts works pretty well with single file components, so I merged this pull request now :) I think we can agree that vues SFC pattern is really awesome 👍 so I'm glad we managed to avoid splitting the component into multiple files.

I have easter vacation all week btw, so if I am active in the middle of the day, that's the reason ;)

**Comment by pingwiniasty** [comment]
Ok, I've just come back from work :)

**Comment by jarlah** [comment]
@pingwiniasty i have created a simple todo app with the same pattern as in this project at https://github.com/jarlah/todo-vue-typescript. For exploring concepts. Do you like the loose coupling? instead of doing this.$store.dispatch('addTodo', newTodo); (which is very unsafe and multiple things can go wrong) we are calling a typeafe method in a todo module. which is connected to the store. I have worked for a long time as developer, and this pattern is appealing. anyway. I think its the way to go.

Way forward now is creating the gui which must be awesome :) Good thing about vue is that creating gui is very easy. Its as simple as creating vue files, with template and style, and combine them with other vue components. Dont need to code anything, i love it ;)
